inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/injection/ClassInjector.java

```diff
@@ -0,0 +1,224 @@
+        if (loader == null) {
+            //for bootstrap classes the standard injection does not work probably
+            //however it also is not necessary has a reference to the bootstrap loader won't cause a memoryleak anyway
+            return bootstrapChildLoader.defineNewClass(className, byteCode);
+        } else {
+            return ClassDefinitionUtils.defineClass(className, byteCode, neighborClass, loader);
+        }
+    }
+
+    private String getUniqueNameInSamePackage(Class<?> neighborClass) {
+
+        String packagePrefix;
+        if (neighborClass.getClassLoader() == null) {
+            //when injecting to the "bootstrap" we actually inject to #bootstrapChildLoader
+            //therefore, we must not reuse any java.* package names
+            packagePrefix = "rocks.inspectit.oce.injected.";
+        } else {
+            packagePrefix = neighborClass.getName();
+            int lastDot = packagePrefix.lastIndexOf('.');
+            packagePrefix = packagePrefix.substring(0, lastDot + 1);
+            if (packagePrefix.isEmpty()) {
+                throw new RuntimeException("Injection into the default package is not supported!");
+            }
+        }
+
+        String namePrefix = packagePrefix + "inspectitGen$$$";
+        ClassLoader loader = neighborClass.getClassLoader();
+
+        while (true) {
+            String name = namePrefix + Math.abs(nameGenerationRandom.nextLong());
+            try {
+                Class.forName(name, false, loader);
+            } catch (ClassNotFoundException e) {
+                return name;
+            }
+        }
+    }
+
+    private void markClassForReuse(String classStructureIdentifier, Class<?> clazz) {
+        ClassLoader loader = clazz.getClassLoader();
+        orphanClasses.computeIfAbsent(loader, (cl) -> new HashMap<>())
+                .computeIfAbsent(classStructureIdentifier, (cl) -> new LinkedList<>())
+                .add(new WeakReference<>(clazz));
+    }
+
+    private Optional<Class<?>> tryReusingClassInLoader(String classStructureIdentifier, ClassLoader loader) {
+        if (loader == null) {
+            loader = bootstrapChildLoader;
+        }
+        collectOrphanClasses();
+        LinkedList<WeakReference<Class<?>>> orphansOfGivenStructureInClassloader =
+                orphanClasses.getOrDefault(loader, Collections.emptyMap())
+                        .getOrDefault(classStructureIdentifier, new LinkedList<>());
+
+        WeakReference<Class<?>> classToReuse = orphansOfGivenStructureInClassloader.pollFirst();
+        if (classToReuse == null) {
+            return Optional.empty();
+        } else {
+            return Optional.ofNullable(classToReuse.get());
+        }
+    }
+
+    private void collectOrphanClasses() {
+        Reference<? extends InjectedClass<?>> freedRef = unusedInjectedClassesQueue.poll();
+        while (freedRef != null) {
+            val classRef = (InjectedClassReference) freedRef;
```

Comment:
can you not type the queue already as the `<InjectedClassReference>` to avoid this cast, the constructor of the InjectedClassReference can still have same signature?